calcite/src/main/java/org/apache/iotdb/calcite/IoTDBFilter.java

```diff
@@ -0,0 +1,321 @@
+
+public class IoTDBFilter extends Filter implements IoTDBRel {
+
+  private final FilterOperator filterOperator;
+  private final List<String> fieldNames;
+  private List<String> predicates;  // for global predicate
+  Map<String, String> deviceToFilterMap = new LinkedHashMap<>();  // for device predicate
+
+  protected IoTDBFilter(RelOptCluster cluster, RelTraitSet traits, RelNode child, RexNode condition)
+      throws LogicalOptimizeException {
+    super(cluster, traits, child, condition);
+    this.fieldNames = IoTDBRules.IoTDBFieldNames(getRowType());
+    this.filterOperator = getIoTDBOperator(condition);
+    this.predicates = translateWhere(filterOperator);
+
+    // add global predicate to each device if both global and device predicate exist
+    if (!this.predicates.isEmpty() && !this.deviceToFilterMap.isEmpty()) {
+      for (String device : deviceToFilterMap.keySet()) {
+        StringBuilder builder = new StringBuilder(this.deviceToFilterMap.get(device));
+        builder.append(Util.toString(predicates, " OR ", " OR ", ""));
+        this.deviceToFilterMap.put(device, builder.toString());
+      }
+    }
+    assert getConvention() == IoTDBRel.CONVENTION;
+    assert getConvention() == child.getConvention();
+  }
+
+  @Override
+  public RelOptCost computeSelfCost(RelOptPlanner planner,
+      RelMetadataQuery mq) {
+    return super.computeSelfCost(planner, mq).multiplyBy(0.1);
+  }
+
+  @Override
+  public Filter copy(RelTraitSet relTraitSet, RelNode input, RexNode condition) {
+    try {
+      return new IoTDBFilter(getCluster(), traitSet, input, condition);
+    } catch (LogicalOptimizeException e) {
+      throw new AssertionError(e.getMessage());
+    }
+  }
+
+  @Override
+  public void implement(Implementor implementor) {
+    implementor.visitChild(0, getInput());
+    implementor.add(deviceToFilterMap, predicates);
+  }
+
+  /**
+   * Translates {@link RexNode} expressions into IoTDB filter operator.
+   */
+  private FilterOperator getIoTDBOperator(RexNode filter) {
+    switch (filter.getKind()) {
+      case EQUALS:
+        return getBasicOperator(SQLConstant.EQUAL, (RexCall) filter);
+      case NOT_EQUALS:
+        return getBasicOperator(SQLConstant.NOTEQUAL, (RexCall) filter);
+      case GREATER_THAN:
+        return getBasicOperator(SQLConstant.GREATERTHAN, (RexCall) filter);
+      case GREATER_THAN_OR_EQUAL:
+        return getBasicOperator(SQLConstant.GREATERTHANOREQUALTO, (RexCall) filter);
+      case LESS_THAN:
+        return getBasicOperator(SQLConstant.LESSTHAN, (RexCall) filter);
+      case LESS_THAN_OR_EQUAL:
+        return getBasicOperator(SQLConstant.LESSTHANOREQUALTO, (RexCall) filter);
+      case AND:
+        return getBinaryOperator(SQLConstant.KW_AND, ((RexCall) filter).getOperands());
+      case OR:
+        return getBinaryOperator(SQLConstant.KW_OR, ((RexCall) filter).getOperands());
+      default:
+        throw new AssertionError("cannot get IoTDBOperator from " + filter);
```

Comment:
why just throws error not exception ?

Reply:
I just used it for convenience before...Sorry, I've added a new exception for it.